### PR TITLE
feat: add scripts for Safe interaction

### PR DESCRIPTION
**Pull request by madlabman**
Add script for constructing multisend transaction for Gnosis Safe to deploy vesting contracts by the given CSV file with parameters.

### CODE REVIEW

File: input.csv [hunk 1]
```diff
@@ -0,0 +1 @@
+amount,recipient,vesting_duration,vesting_start,cliff_length,is_fully_revokable
```
Comment (dgusakov) [inline input.csv]: input.csv.example

File: pyproject.toml [hunk 2]
```diff
@@ -11,6 +11,7 @@ packages = [{include = "lido_vesting_escrow"}]
 python = ">=3.8,<3.11"
 eth-brownie = "^1.19.2"
 dotmap = "^1.3.30"
+ape-safe = { git = "https://github.com/madlabman/ape-safe", tag = "v0.6.1" }
```
Comment (dgusakov) [inline pyproject.toml]: Pls submit PR with fixes to the upstream

File: scripts/build_multisig_tx.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            params.is_fully_revokable,
+            {"from": safe.address},
+        )
+        deployments.append((params, tx))
+
+    log.info("Validate deployed vestings")
+    params: VestingParams
+    tx: TransactionReceipt
+    for params, tx in deployments:
+        log.info(f"Checking {params.recipient} vesting")
```
Comment (dgusakov) [inline scripts/build_multisig_tx.py]:
```suggestion
        log.info(f"Checking {tx.return_value} vesting for recipient {params.recipient}")
```

File: scripts/safe_delegates.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+
+if __name__ == "__main__":  # module-like invocation
+    try:
+        _, cmd, *args = sys.argv
+    except ValueError:
+        log.error("Usage: python -m scripts.safe_delegates cmd ...args")
+        _abort_execution()
+
+    if cmd not in vars():
```
Comment (dgusakov) [inline scripts/safe_delegates.py]: What does` vars()` func stands for?

File: scripts/build_multisig_tx.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    factory = VestingEscrowFactory.at(
+        address=config["FACTORY_ADDRESS"],
+        owner=safe.address,
+    )
+
+    # Uncomment for testing purpose
+    # factory = VestingEscrowFactory.at(
+    #     _test_factory_address(safe),
+    #     owner=safe.address,
+    # )
```
Comment (dgusakov) [inline scripts/build_multisig_tx.py]: Use flag instead of commented code for testing mode

**Review by dgusakov — CHANGES_REQUESTED** [review]

File: scripts/safe_delegates.py (hunk 4 above)
Comment (madlabman) [reply scripts/safe_delegates.py]: It acts as a `locals()` and returns local symbols table, actually should be better to change to `locals`, thx.

File: scripts/build_multisig_tx.py (hunk 5 above)
Comment (madlabman) [reply scripts/build_multisig_tx.py]: Brownie's scripts runner doesn't support providing flags, added as argument instead.

**Comment by dgusakov** [comment]
Also add info to the README

File: input.csv (hunk 1 above)
Comment (madlabman) [reply input.csv]: Done

**Review by dgusakov — APPROVED** [review]

File: scripts/multisig_tx.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+LDO_ADDRESS = "0x5A98FcBEA516Cf06857215779Fd812CA3beF1B32"
+
+
+def build(csv_filename: str, non_empty_for_prod=None):
+    """Build vesting contracts deployment tx by parameters defined in CSV file"""
+    is_prod = bool(non_empty_for_prod)
+    if is_prod:
+        log.warn("SCRIPT RUNNED IN PRODUCTION ENV")
+        if not log.prompt_yes_no("ARE YOU SURE TO CONTINUE?"):
```
Comment (dgusakov) [inline scripts/multisig_tx.py]:
```suggestion
        if not log.prompt_yes_no("ARE YOU SURE YOU WANT TO CONTINUE?"):
```

File: scripts/multisig_tx.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+                params.vesting_start,
+                params.cliff_length,
+                params.is_fully_revokable,
+                {"from": safe.address},
+            )
+        safe_tx = safe.multisend_from_receipts()
+
+    _preview_and_check_tx(safe, safe_tx, params_list, is_prod)
+
+    if log.prompt_yes_no("Sign with frame?"):
```
Comment (dgusakov) [inline scripts/multisig_tx.py:81]: Should we also add manual signing? frame is still in beta

File: scripts/multisig_tx.py [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+    params_list = [VestingParams.from_tuple(p) for p in raw_params_list]
+
+    safe = ApeSafe(config["SAFE_ADDRESS"])
+
+    log.info("Retrieving transaction from Gnosis Safe")
+    safe_tx = safe.get_safe_tx_by_safe_tx_hash(safe_tx_hash)
+    _preview_and_check_tx(safe, safe_tx, params_list, is_prod=True)
+
+
+def sign(safe_tx_hash: str) -> None:
```
Comment (dgusakov) [inline scripts/multisig_tx.py]: Unused method

File: scripts/multisig_tx.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+    input("Press ENTER to exit...")
+
+
+def _preview_and_check_tx(safe: ApeSafe, safe_tx: SafeTx, params_list: Sequence["VestingParams"], is_prod=False):
+    vestings_sum = sum(p.amount for p in params_list)
+    starting_balance = _ldo_balance(safe.address)
+
+    log.info("Preview transaction")
+    # do not reset chain in testing to keep the fake factory and other contracts intact
+    tx = safe.preview(safe_tx, reset=is_prod)
```
Comment (dgusakov) [inline scripts/multisig_tx.py]: Prompt after preview?

File: scripts/multisig_tx.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    return ldo.balanceOf(account)
+
+
+def _check_tx(tx: TransactionReceipt, params_list: Sequence[VestingParams]) -> None:
+    """Check contracts created by the transaction against the parsed vesting parameters"""
+    log.info("Validate contracts from multisend transaction")
+
+    new_contracts_count = len(tx.new_contracts) if tx.new_contracts else 0
+    if len(params_list) != new_contracts_count:
+        raise RuntimeError("Deployed contracts count mismatch")
```
Comment (dgusakov) [inline scripts/multisig_tx.py]:
```suggestion
        raise RuntimeError(f"Deployed contracts count mismatch. Expected: {len(params_list)}. Actual: {new_contracts_count}")
```

File: scripts/multisig_tx.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+
+    address: str
+    for address in tx.new_contracts:
+        contract = VestingEscrow.at(address)
+        recipient = contract.recipient()
+        try:
+            [params] = [p for p in params_list if p.recipient == recipient]
+        except ValueError as e:
+            raise ValueError(f"Recipient {recipient} not found in source") from e
```
Comment (dgusakov) [inline scripts/multisig_tx.py]:
```suggestion
            raise ValueError(f"Recipient {recipient} for {address=} was not found in source") from e
```

File: scripts/multisig_tx.py (hunk 8 above)
Comment (madlabman) [reply scripts/multisig_tx.py]: It is a command.

File: scripts/multisig_tx.py (hunk 9 above)
Comment (madlabman) [reply scripts/multisig_tx.py]: It will be right after the balance check, which we want to fail fast.

File: scripts/multisig_tx.py (hunk 7 above)
Comment (madlabman) [reply scripts/multisig_tx.py:81]: Added function for this. It is not very detailed since it rare case.

**Review by dgusakov — APPROVED** [review]

File: scripts/multisig_tx.py [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+        address=factory_address,
+        owner=safe.address,
+    )
+
+    log.info("Checking multisig balance")
+    starting_balance = _ldo_balance(safe.address)
+    vestings_sum = sum(p.amount for p in params_list)
+    assert starting_balance >= vestings_sum, f"Not enough LDOs for vesting, need at least {vestings_sum / 10 ** 18}"
+
+    log.info("Constructing multisend transaction")
```
Comment (dgusakov) [inline scripts/multisig_tx.py]: We need to add multisig `approve` TX with the sum of the LDO required for vestings
OR
We can approve infinite LDO once. Pls ask @kadmil

File: pyproject.toml (hunk 2 above)
Comment (madlabman) [reply pyproject.toml]: There are actually 2 or 3 PRs open with patches included in the fork now. By the time of writing, these PRs are still actual but have no response.

File: scripts/multisig_tx.py (hunk 12 above)
Comment (madlabman) [reply scripts/multisig_tx.py]: Added LDO approve to the sum of vesting under deploy.

**Review by dgusakov — APPROVED** [review]

**Review by dgusakov — APPROVED** [review]

**Pull request merged**